include/uk/plat/mte.h

```diff
@@ -0,0 +1,42 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
```

Comment:
Well, since it should be initialized by the platform I envisioned support is hierarchical, as arch -> plat. Exporting this via `include/uk/plat` would allow applications to tag memory even without libc. But I did that for convenience, rather than having a use-case in mind, so I'm happy to revert and restrict to internal use only. Saying that, in the Unikraft gathering we discussed a bit the possibility to deprecate the requirement of building for multiple platforms on a single build and allow applications to access internal platform / arch functionality (still inconclusive). The prospect of this being supported would also be valid argumentation towards making `include/uk/plat/mte.h` redundant. Last but not least, applications can always manually include internal headers in their makefiles.

Also, now that we will soon have POSIX `mmap()`, we can instead add support for the `PROT_MTE` there. I can do that after #603 is merged.

> But then I also think selling MTE as a broadly available feature on different architectures by placing the header globally in uk/arch is not really optimal either. Especially, if at the same time, we add checks for ARM_FEAT_MTE everywhere (e.g., ukalloc).

Another possibility is to give this feature a more generic name like `ukarch_memtag`, given that most likely such a feature will eventually also land in x86_64 (perhaps it already exists there too?), and for now, as you suggest, stub it out there. Then eliminate the various `#ifdef`s. What do you think?

Reply:
It does not really fit into `include/uk/plat` because the header includes an `asm` path, which only architectural headers should do. And it makes sense to have the actual functions defined in `asm` because this is an architectural feature and not a platform feature.

However, all functionality defined in `include/uk/[arch|plat]` should use the `uk_`, `ukarch_` or `ukplat_` prefixes. So, yes. Strictly speaking, this needs to get a name like `ukarch_memtag` with stubs for x86. It is probably the cleaner solution even in the case there won't be a x86 memory tagging feature.

Reply:
@marcrittinghaus agreed. Notice that introducing a generic `ukarch_memtag` API requires making some assumptions on how memory tagging will be implemented in other architectures. I am deriving the requirements of the API from what we have needed so far for ukalloc:
1. A way to tag regions
2. A way to align region size to tag granularity
3. A way to strip the tag from a pointer

Point 1 is quite generic and I think it is safe to assume that a function such as `void *ukarch_memtag_region(void *ptr, __sz size);` will do. Notice however that even that function returns a tagged pointer, which makes an assumption about the implementation, that is tags are stored in pointers. This assumption is carried over to the 3rd point. For point 2 we also have an assumption about granularity requirements, but it's easy to overcome if some architecture does not impose any alignment requirements. Perhaps all these are a bit stretched, and in practice it's likely that all implementations will look the same, more or less. But still, before we introduce this interface, perhaps it's worth to be aware of the risks.

So based on the above, I see the following options:
1. Introduce a generic `ukarch_memtag` with the above features, based on the assumption that implementations in other architectures will look similar.
2. Move the public interface of MTE from `include/uk/plat/` to `include/uk/arch/` and reconsider generalizing if other architectures introduce a similar feature.

What do you think?

Reply:
@michpappas These are valid points and perhaps it is really better to wait until we see other architectures implementing this to come up with a safe generic interface. However, I am not super happy because it would be the first header there that introduces functions without the `ukarch_` prefix and which are only available on a certain architecture. It also means that the changes in the allocator(s) are ARM specific and need `#ifdef` guards everywhere. Also, for compatibility, would we need to keep this ARM-specific interface even if we introduce a generic interface?

So, maybe in the end it does not make much of a difference if we would (change)/extend an existing API vs. introducing one later and (removing)/keeping the old ARM-specific one around.

Could you imagine a different way to tag regions besides encoding it in the pointer?

From my perspective, we should move this to `include/uk/arch` anyways.

Reply:
@marcrittinghaus 
> However, I am not super happy because it would be the first header there that introduces functions without the ukarch_ prefix and which are only available on a certain architecture.

Ah yes, I forgot about that.

> Could you imagine a different way to tag regions besides encoding it in the pointer?

I cannot, that's why I said all this are probably a bit stretched 🙂 

Okay, then I will update the PR and introduce the generic `ukarch_memtag` interface, thanks for the feedback 👍🏼